Add bug scrub

cpaelzer opened the pull request:
This improves the normal use-case output.

From
```
$ ustriage
Ubuntu Server Bug List
Please be patient, this can take a few minutes...
Team 'ubuntu-server' currently subscribed to 420 bugs
No date set, auto-search yesterday/weekend for the most common triage.
---
'*': ubuntu-server is directly subscribed
'+': last bug activity is ours
Bugs last updated between 2021-09-17 (Friday) and 2021-09-19 (Sunday) inclusive
Date range identified as: "Monday triage"
Found 22 bugs
LP: #1828074 - *+(Triaged)      [augeas]         - Memory usage issues with large Apache httpd configuration files
LP: #1885403 - *(Fix Committed) [postfix]        - posttls-finger fails to connect to private/tlsmgr
LP: #1888345 - *(Invalid)       [sysstat]        - Incorrect values  on iostat command when using JSON output
LP: #1892292 - *(Invalid)       [open-iscsi]     - iscsi boot fails to boot into rootfs
LP: #1917306 - (Incomplete)     [qemu]           - [20.04 FEAT] zKVM: Add support for virtio-fs - qemu (backport)
LP: #1930398 - (Incomplete)     [libvirt]        - libvirtd  unable to execute QEMU command 'blockdev-remove-medium'
LP: #1932175 - (Fix Released)   [qemu]           - [21.10 FEAT] CPU Model for new IBM Z Hardware - qemu part
LP: #1936744 - (Expired)        [apache2]        - Network adapter disappeared.Disconnected from internet.
LP: #1936761 - (Expired)        [clamav]         - package clamav-base 0.103.2+dfsg-0ubuntu0.20.04.2 failed to install/upgrade: installed clamav-base package post-installation script subprocess returned error exit status 1
LP: #1936860 - (Expired)        [spamassassin]   - package sa-compile 3.4.4-1ubuntu1.1 failed to install/upgrade: installed sa-compile package post-installation script subprocess returned error exit status 1
LP: #1938144 - (Triaged)        [openssh]        - monitor_read: unpermitted request 48 on server while attempting GSSAPI key exchange
LP: #1938897 - (Fix Released)   [golang]         - snapd 2.51.1+21.10 autotest segfault on ppc64
LP: #1939352 - (Fix Released)   [squid]          - FTBFS 4.13-10ubuntu3 when building with GCC 11
LP: #1940079 - *(Incomplete)    [strongswan]     - Strongswan doesn't support TPM 2.0 through the TSS2 interface
LP: #1940603 - *(Invalid)       [postfix]        - postfix/smtpd: fatal: no SASL authentication
LP: #1943823 - (New)            [ruby2.7]        - ruby2.7 ftbfs on ppc64el using GCC 11.2
LP: #1943847 - (Incomplete)     [libvirt]        - virsh migrate with LVM volume as disk fails
LP: #1943989 - (New)            [dovecot]        - package dovecot-core 1:2.3.13+dfsg1-1ubuntu1 failed to install/upgrade: installed dovecot-core package post-installation script subprocess returned error exit status 1
LP: #1943994 - (New)            [samba]          - package samba-common-bin 2:4.13.5+dfsg-2ubuntu2 failed to install/upgrade: installed samba-common-bin package post-installation script subprocess returned error exit status 1
LP: #1944005 - (New)            [multipath-tools] - System Lockup during failover
LP: #1944008 - (New)            [php8.0]         - php-fpm doesn't include huge-code-pages feature support
LP: #1944037 - (New)            [liburing]       - Latest packaged version is a year old with a lot of bug fixes missing

---
Bugs tagged 'server-next' and not touched in 60 days
Found 0 bugs

---
Bugs in backlog and not touched in 180 days
Found 2 bugs
LP: #1919923 - (Confirmed)      [libapache2-mod-python] - mod_python statically linked to wrong python 2.7 version
LP: #1919965 - (Triaged)        [nginx]          - nginx-common nginx.service comments use wrong signal
```

to

```
$ python -m ustriage 
Ubuntu Server Triage helper
Please be patient, this can take a few minutes...
Team 'ubuntu-server' currently subscribed to 420 bugs
No date set, auto-search yesterday/weekend for the most common triage.
---
'*': ubuntu-server is directly subscribed
'+': last bug activity is ours
Bugs last updated between 2021-09-17 (Friday) and 2021-09-19 (Sunday) inclusive
Date range identified as: "Monday triage"
Found 22 bugs
LP: #1828074 - *+(Triaged)      17.09.21 [augeas]         => mirespace  - Memory usage issues with large Apache httpd configuration files
LP: #1885403 - *(Fix Committed) 19.09.21 [postfix]        => mirespace  - posttls-finger fails to connect to private/tlsmgr
LP: #1888345 - *(Invalid)       17.09.21 [sysstat]        => mirespace  - Incorrect values  on iostat command when using JSON output
LP: #1892292 - *(Invalid)       17.09.21 [open-iscsi]                   - iscsi boot fails to boot into rootfs
LP: #1917306 - (Incomplete)     17.09.21 [qemu]           => canonical  - [20.04 FEAT] zKVM: Add support for virtio-fs - qemu (backport)
LP: #1930398 - (Incomplete)     18.09.21 [libvirt]                      - libvirtd  unable to execute QEMU command 'blockdev-remove-medium'
LP: #1932175 - (Fix Released)   17.09.21 [qemu]           => canonical  - [21.10 FEAT] CPU Model for new IBM Z Hardware - qemu part
LP: #1936744 - (Expired)        18.09.21 [apache2]                      - Network adapter disappeared.Disconnected from internet.
LP: #1936761 - (Expired)        18.09.21 [clamav]                       - package clamav-base 0.103.2+dfsg-0ubuntu0.20.04.2 failed to install/upgrade: installed clamav-base package post-installation script subprocess returned error exit status 1
LP: #1936860 - (Expired)        19.09.21 [spamassassin]                 - package sa-compile 3.4.4-1ubuntu1.1 failed to install/upgrade: installed sa-compile package post-installation script subprocess returned error exit status 1
LP: #1938144 - (Triaged)        17.09.21 [openssh]                      - monitor_read: unpermitted request 48 on server while attempting GSSAPI key exchange
LP: #1938897 - (Fix Released)   17.09.21 [golang]                       - snapd 2.51.1+21.10 autotest segfault on ppc64
LP: #1939352 - (Fix Released)   18.09.21 [squid]          => sergiodj   - FTBFS 4.13-10ubuntu3 when building with GCC 11
LP: #1940079 - *(Incomplete)    17.09.21 [strongswan]     => paride     - Strongswan doesn't support TPM 2.0 through the TSS2 interface
LP: #1940603 - *(Invalid)       19.09.21 [postfix]                      - postfix/smtpd: fatal: no SASL authentication
LP: #1943823 - (New)            17.09.21 [ruby2.7]                      - ruby2.7 ftbfs on ppc64el using GCC 11.2
LP: #1943847 - (Incomplete)     17.09.21 [libvirt]                      - virsh migrate with LVM volume as disk fails
LP: #1943989 - (New)            17.09.21 [dovecot]                      - package dovecot-core 1:2.3.13+dfsg1-1ubuntu1 failed to install/upgrade: installed dovecot-core package post-installation script subprocess returned error exit status 1
LP: #1943994 - (New)            17.09.21 [samba]                        - package samba-common-bin 2:4.13.5+dfsg-2ubuntu2 failed to install/upgrade: installed samba-common-bin package post-installation script subprocess returned error exit status 1
LP: #1944005 - (New)            17.09.21 [multipath-too]                - System Lockup during failover
LP: #1944008 - (New)            17.09.21 [php8.0]                       - php-fpm doesn't include huge-code-pages feature support
LP: #1944037 - (New)            18.09.21 [liburing]                     - Latest packaged version is a year old with a lot of bug fixes missing

---
Bugs tagged "server-next" and not touched in 60 days
Found 0 bugs

---
Bugs in backlog and not touched in 180 days
Found 2 bugs
LP: #1919923 - (Confirmed)      23.03.21 [libapache2-mo]                - mod_python statically linked to wrong python 2.7 version
LP: #1919965 - (Triaged)        22.03.21 [nginx]          => teward     - nginx-common nginx.service comments use wrong signal
```

UI wise that is

-  avoid broken layout by long pkg names and such
- report assignee
- report last updated date

And it brings a new-feature by re-using the code (a lot would have been copy and paste) for the bug-scrub which reports

- all server-next bugs
- all server subscribed backlog
- one can limit the length of the latter

An example of that is here:
```
$ python -m ustriage --bug-scrub --limit-backlog 20
Ubuntu Server Triage helper
Please be patient, this can take a few minutes...

---
Bugs tagged "server-next"
Found 15 bugs
LP: #1719354 - +(Triaged)       14.07.20 [samba]          => ahasenack  - apparmor blocking smbd which is in complain mode
LP: #1778073 - +(Incomplete)    15.09.21 [dnsmasq]        => paride     - dnsmasq and resolvconf hangs on start
LP: #1872145 - +(Triaged)       30.04.20 [openssh]                      - explicit key offered after all agent keys, auth can fail before explicit key used
LP: #1913421 - +(In Progress)   14.09.21 [qemu]                         - Load of pre-upgrade qemu modules needs to avoid noexec
LP: #1412548 - (Triaged)        19.10.17 [resource-agen]                - OCF Resource Agent Filesystem is still CMAN dependent
LP: #1439649 - (Confirmed)      30.04.20 [pacemaker]                    - Pacemaker unable to communicate with corosync on restart under lxc
LP: #1785383 - (Triaged)        06.09.21 [dnsmasq]                      - missing EDNS0 record confuses systemd-resolved
LP: #1832182 - (Fix Committed)  28.07.21 [apache2]        => bryce      - systemd unable to detect running apache if invoked via apache2ctl graceful
LP: #1882527 - (Triaged)        13.09.21 [mysql-5.7]      => paride     - mysql timeoutsec results in killing mysql process
also:        - (Fix Committed)           [mysql-8.0]      => paride    
LP: #1885403 - (Fix Committed)  19.09.21 [postfix]        => mirespace  - posttls-finger fails to connect to private/tlsmgr
LP: #1912118 - (Incomplete)     08.05.21 [dovecot]        => bryce      - assertion failure in message_part_finish when searching large folder
LP: #1925182 - (Triaged)        15.09.21 [clamav]                       - ERROR: Can't open /var/log/clamav/freshclam.log in append mode (check permissions!)
LP: #1936894 - (Triaged)        04.09.21 [qemu]                         - microvm is not the default type for qemu-system-x86_64-microvm >=qemu5.2
LP: #1939292 - (Triaged)        20.09.21 [dovecot]        => utkarsh    - Error when attempting search on server, fix already upstream

---
Bugs in backlog
Found 471 bugs
Displaying top & bottom 20
# Recent tasks #
Found 20 bugs
LP: #1828074 - (Triaged)        17.09.21 [augeas]         => mirespace  - Memory usage issues with large Apache httpd configuration files
LP: #1916485 - (New)            15.09.21 [libseccomp]                   - test -x fails inside shell scripts in containers
also:        - (New)                     [docker.io]                   
also:        - (New)                     [runc]                        
LP: #1933143 - (Fix Committed)  15.09.21 [open-vm-tools]                - Open-vm-tools release 11.3.0 has been released
LP: #1940913 - (Fix Committed)  13.09.21 [dpdk]                         - MRE updates 19.11.10 and 20.11.3
LP: #1878601 - (Triaged)        09.09.21 [nfs-utils]                    - nfs-utils in Debian and Ubuntu is *very* old
LP: #1864612 - (Triaged)        07.09.21 [ipmitool]                     - requests ipmi-tool to include lasted patch for supporting quanta server
LP: #1851232 - (Triaged)        06.09.21 [open-iscsi]                   - Laptop does not suspend when using open-iscsi
LP: #1572132 - (Confirmed)      04.09.21 [linux]                        - SMBv2+ does not show all files when mounting a windows share
LP: #1942195 - (Confirmed)      02.09.21 [samba]                        - Installing Samba unexpectedly adds many unknown local users to sambashare group
LP: #1753329 - (Triaged)        31.08.21 [docker.io]                    - deprecate and stop packaging?
LP: #1505670 - (New)            31.08.21 [squid-deb-pro]                - uncaptured python exception
also:        - (New)                     [python2.7]                   
LP: #1933378 - (Confirmed)      27.08.21 [mongodb]                      - Unable to build from source mongodb-server-core - focal
LP: #1865900 - (In Progress)    26.08.21 [apache2]        => mdeslaur   - apache 2.4.29-1ubuntu4.12 authentication with client certificate broken
also:        - (Confirmed)               [requests]                    
also:        - (Confirmed)               [python-urllib]               
LP: #1940916 - (Triaged)        25.08.21 [monitoring-pl]                - Incorrectly excludes tmpfs filesystems
LP: #1882272 - (New)            25.08.21 [lm-sensors]                   - fancontrol does not work after sleep/wakeup
---------------------------------------------------
# Oldest tasks #
Found 20 bugs
LP: #1511222 - (Triaged)        02.08.16 [apache2]        => wesley-wi  - Incorrect trusted proxy match test in mod_remoteip
LP: #1524635 - (Triaged)        14.06.16 [haproxy]                      - haproxy syslog configuration causes double logging
LP: #1428288 - (Triaged)        31.03.16 [apache2]                      - dist-upgrade failed: apache2 postinst failed on absolute symlinks in mods-enabled
LP: #1403499 - (Triaged)        31.03.16 [apache2]                      - apache2 mod_reqtimeout erroneous timeouts
LP: #1561007 - (New)            24.03.16 [squid3]                       - Upstream Bug #3769: client_netmask not evaluated since Comm redesign
LP: #1375821 - (Triaged)        03.03.16 [ifupdown]                     - ifupdown doesn't bring loopback to state up
LP: #1267564 - (Confirmed)      21.10.15 [etckeeper]                    - etckeeper with git breaks package upgrades that touch /etc
LP: #1505556 - (Triaged)        16.10.15 [munin]                        - lpstat plugin doesn't return job count
LP: #1335391 - (Confirmed)      23.09.15 [etckeeper]                    - etckeeper with git breaks update-manager
LP: #1492212 - (Triaged)        04.09.15 [bind9]                        - postinst should validate config before restarting bind
LP: #1481337 - (Confirmed)      01.09.15 [keepalived]                   - keepalived makes a floating IP available on more than one host after configuration reload
LP: #1481579 - (New)            05.08.15 [amavisd-new]                  - amavisd-new throws stderr when running cron job
LP: #1242376 - (Triaged)        27.07.15 [php5]                         - reload php5-fpm crashes
LP: #1428796 - (New)            06.03.15 [apache2]                      - service stop; service start fails with wsgi.py from freeipa-server
LP: #609625  - (Confirmed)      14.05.11 [linux]                        - kexec-tools 0_kdump
LP: #28706   - (Confirmed)      17.08.10 [nfs-utils]                    - need way to specify the lockd port
LP: #150649  - (Confirmed)      14.06.10 [libapache-mod]                - gutsy does not have a working apache+mysql authentication solution
LP: #338047  - (Triaged)        12.03.10 [redhat-cluste]                - soft lockup - CPU#0 stuck for 11s! [clurgmgrd:4587]
LP: #222620  - (Confirmed)      10.08.08 [ebox-ntp]                     - package ebox-ntp None failed to install/upgrade: subprocess post-removal script returned error exit status 1
LP: #214863  - (Confirmed)      09.04.08 [nagios-plugin]                - various bugs and trust issues
```

paride commented:
Thanks @cpaelzer for implementing this! It would be nice to have some basic CI for this repository, so I proposed https://github.com/canonical/ubuntu-server-triage/pull/29, which fixes the tox run for `master` and adds a GH Actiona that runs tox on pushes and on PR.

There would be more to do:
 - Rename master to main
 - Update snapcraft.yaml so it point to this repo and not to the original one from Josh
 - Ensure that the snap actually updates from here

But let's start from this!

paride commented:
My proposal is to first land https://github.com/canonical/ubuntu-server-triage/pull/29 and then rebase this branch.

paride reviewed (changes requested):
My proposal is to first land #29 and then rebase this branch.

cpaelzer commented:
FYI tox is happy now, snapcraft build is impreoved but not yet complete again.

cpaelzer commented:
> My proposal is to first land #29 and then rebase this branch.

This happened but blocks this via "change-requested", you might remove this CR and if you have time re-review?

paride reviewed (commented):
In general this LGTM. I tried to `snapcraft build` but it failed with:
```
Building ustriage 
+ snapcraftctl build
+ python3 -m venv /root/parts/ustriage/install
Error: Command '['/root/parts/ustriage/install/bin/python3', '-Im', 'ensurepip', '--upgrade', '--default-pip']' returned non-zero exit status 1.
Failed to build 'ustriage'.
```
Is this the same failure you're seeing?

cpaelzer commented:
> + python3 -m venv /root/parts/ustriage/install
> Error: Command '['/root/parts/ustriage/install/bin/python3', '-Im', 'ensurepip', '--upgrade', '--default-pip']' returned non-zero exit status 1.
> Failed to build 'ustriage'.
> ```
> 
> Is this the same failure you're seeing?

Yes that is due to core20, it works fine with 18 which I'll adapt in a bit.

cpaelzer commented:
Discussed with Paride, he preferred (and I'm ok) a more selective approach.
So I'll (in testing atm) add a change to convert --bug-scrub into

-   -T, --show-tagged     Display (--tag-next or default) tagged bugs
-   -B, --show-backlog    Display backlog of all (--lpname or default) subscribed bugs
-   -N, --show-no-triage  Do not Display the default triage content (recent and expiring bugs).

cpaelzer commented:
As I said this opened up for more use-cases that were so close we just had to add/open-up for them.
I'll post how the various example outputs now look like and push my branch.

Usual triage usage
```
$ ustriage
Ubuntu Server Triage helper
Please be patient, this can take a few minutes...
Team 'ubuntu-server' currently subscribed to 422 bugs
No date set, auto-search yesterday/weekend for the most common triage.
---
'*': ubuntu-server is directly subscribed
'+': last bug activity is ours
Bugs last updated on 2021-09-20 (Monday)
Date range identified as: "Tuesday triage"
Found 21 bugs
LP: #1749393 - +(Triaged)       20.09.21 [qemu]                         - sbrk() not working under qemu-user with a PIE-compiled binary?
LP: #1872145 - *+(Triaged)      20.09.21 [openssh]                      - explicit key offered after all agent keys, auth can fail before explicit key used
LP: #1912118 - *+(In Progress)  20.09.21 [dovecot]        => bryce      - assertion failure in message_part_finish when searching large folder
LP: #1919923 - *+(Confirmed)    20.09.21 [libapache2-mo]                - mod_python statically linked to wrong python 2.7 version
LP: #1919965 - *+(Triaged)      20.09.21 [nginx]          => teward     - nginx-common nginx.service comments use wrong signal
LP: #1943989 - +(Incomplete)    20.09.21 [dovecot]                      - package dovecot-core 1:2.3.13+dfsg1-1ubuntu1 failed to install/upgrade: installed dovecot-core package post-installation script subprocess returned error exit status 1
LP: #1943994 - +(Incomplete)    20.09.21 [samba]                        - package samba-common-bin 2:4.13.5+dfsg-2ubuntu2 failed to install/upgrade: installed samba-common-bin package post-installation script subprocess returned error exit status 1
LP: #1944005 - +(Incomplete)    20.09.21 [multipath-too]                - System Lockup during failover
LP: #1944008 - +(Triaged)       20.09.21 [php8.0]                       - php-fpm doesn't include huge-code-pages feature support
LP: #1944037 - *+(Triaged)      20.09.21 [liburing]                     - Latest packaged version is a year old with a lot of bug fixes missing
LP: #1944104 - +(Incomplete)    20.09.21 [qemu]                         - nested vm on vm not running
LP: #1933143 - *(Fix Released)  20.09.21 [open-vm-tools]                - Open-vm-tools release 11.3.0 has been released
LP: #1938144 - (Triaged)        20.09.21 [openssh]                      - monitor_read: unpermitted request 48 on server while attempting GSSAPI key exchange
LP: #1941745 - (New)            20.09.21 [sosreport]                    - [FFe][sync][sru] sos upstream 4.2
LP: #1941799 - (In Progress)    20.09.21 [openssh]        => jawn-smit  - Please merge openssh 1:8.4p1-6 from Debian unstable
LP: #1943447 - (Fix Released)   20.09.21 [golang]         => bugproxy   - Go runtime: ppc64x binaries randomly segfault on linux 5.13rc6 (#46803)
LP: #1943536 - *(Invalid)       20.09.21 [squid]                        - Number filedescriptors allowed too low for squid on Focal
LP: #1943823 - *(Triaged)       20.09.21 [ruby2.7]                      - ruby2.7 ftbfs on ppc64el using GCC 11.2
LP: #1944106 - (New)            20.09.21 [php7.4]                       - package libapache2-mod-php7.4 7.4.3-4ubuntu2.6 failed to install/upgrade: installed libapache2-mod-php7.4 package post-installation script subprocess returned error exit status 1
LP: #304393  - (Fix Released)   20.09.21 [rpcbind]        => ddstreet   - rpcbind grabs ports used by other daemons such as cupsd
LP: #720638  - (Incomplete)     20.09.21 [net-snmp]                     - snmpd crashes with segfault (libnetsnmpmibs.so.15.1.2)

---
Bugs tagged "server-next" and subscribed "ubuntu-server" and not touched in 60 days
Found 0 bugs

---
Bugs subscribed to ubuntu-server and not touched in 180 days
Found 0 bugs
```

Intended bug-scrub usage
```
ustriage --show-no-triage --show-tagged --show-subscribed --limit-subscribed 20
Ubuntu Server Triage helper
Please be patient, this can take a few minutes...

---
Bugs tagged "server-next" and subscribed "ubuntu-server"
Found 13 bugs
LP: #1719354 - +(Triaged)       14.07.20 [samba]          => ahasenack  - apparmor blocking smbd which is in complain mode
LP: #1778073 - +(Incomplete)    15.09.21 [dnsmasq]        => paride     - dnsmasq and resolvconf hangs on start
LP: #1912118 - +(In Progress)   20.09.21 [dovecot]        => bryce      - assertion failure in message_part_finish when searching large folder
LP: #1913421 - +(In Progress)   14.09.21 [qemu]                         - Load of pre-upgrade qemu modules needs to avoid noexec
LP: #1412548 - (Triaged)        19.10.17 [resource-agen]                - OCF Resource Agent Filesystem is still CMAN dependent
LP: #1439649 - (Confirmed)      30.04.20 [pacemaker]                    - Pacemaker unable to communicate with corosync on restart under lxc
LP: #1785383 - (Triaged)        06.09.21 [dnsmasq]                      - missing EDNS0 record confuses systemd-resolved
LP: #1832182 - (Fix Committed)  28.07.21 [apache2]        => bryce      - systemd unable to detect running apache if invoked via apache2ctl graceful
LP: #1882527 - (Triaged)        13.09.21 [mysql-5.7]      => paride     - mysql timeoutsec results in killing mysql process
also:        - (Fix Committed)           [mysql-8.0]      => paride    
LP: #1885403 - (Fix Committed)  19.09.21 [postfix]        => mirespace  - posttls-finger fails to connect to private/tlsmgr
LP: #1925182 - (Triaged)        15.09.21 [clamav]                       - ERROR: Can't open /var/log/clamav/freshclam.log in append mode (check permissions!)
LP: #1936894 - (Triaged)        04.09.21 [qemu]                         - microvm is not the default type for qemu-system-x86_64-microvm >=qemu5.2

---
Bugs subscribed to ubuntu-server
Found 474 bugs
Displaying top & bottom 20
# Recent tasks #
Found 20 bugs
LP: #1828074 - (In Progress)    21.09.21 [augeas]         => mirespace  - Memory usage issues with large Apache httpd configuration files
LP: #1940079 - (In Progress)    21.09.21 [strongswan]     => paride     - Strongswan doesn't support TPM 2.0 through the TSS2 interface
LP: #1919965 - (Triaged)        20.09.21 [nginx]          => teward     - nginx-common nginx.service comments use wrong signal
LP: #1919923 - (Confirmed)      20.09.21 [libapache2-mo]                - mod_python statically linked to wrong python 2.7 version
LP: #1944037 - (Triaged)        20.09.21 [liburing]                     - Latest packaged version is a year old with a lot of bug fixes missing
LP: #1943823 - (Triaged)        20.09.21 [ruby2.7]                      - ruby2.7 ftbfs on ppc64el using GCC 11.2
LP: #1872145 - (Triaged)        20.09.21 [openssh]                      - explicit key offered after all agent keys, auth can fail before explicit key used
LP: #1916485 - (New)            15.09.21 [docker.io]                    - test -x fails inside shell scripts in containers
also:        - (New)                     [libseccomp]                  
also:        - (New)                     [runc]                        
LP: #1940913 - (Fix Committed)  13.09.21 [dpdk]                         - MRE updates 19.11.10 and 20.11.3
LP: #1878601 - (Triaged)        09.09.21 [nfs-utils]                    - nfs-utils in Debian and Ubuntu is *very* old
LP: #1864612 - (Triaged)        07.09.21 [ipmitool]                     - requests ipmi-tool to include lasted patch for supporting quanta server
LP: #1851232 - (Triaged)        06.09.21 [open-iscsi]                   - Laptop does not suspend when using open-iscsi
LP: #1572132 - (Confirmed)      04.09.21 [linux]                        - SMBv2+ does not show all files when mounting a windows share
LP: #1942195 - (Confirmed)      02.09.21 [samba]                        - Installing Samba unexpectedly adds many unknown local users to sambashare group
LP: #1753329 - (Triaged)        31.08.21 [docker.io]                    - deprecate and stop packaging?
LP: #1505670 - (New)            31.08.21 [squid-deb-pro]                - uncaptured python exception
also:        - (New)                     [python2.7]                   
LP: #1933378 - (Confirmed)      27.08.21 [mongodb]                      - Unable to build from source mongodb-server-core - focal
---------------------------------------------------
# Oldest tasks #
Found 20 bugs
LP: #1511222 - (Triaged)        02.08.16 [apache2]        => wesley-wi  - Incorrect trusted proxy match test in mod_remoteip
LP: #1524635 - (Triaged)        14.06.16 [haproxy]                      - haproxy syslog configuration causes double logging
LP: #1428288 - (Triaged)        31.03.16 [apache2]                      - dist-upgrade failed: apache2 postinst failed on absolute symlinks in mods-enabled
LP: #1403499 - (Triaged)        31.03.16 [apache2]                      - apache2 mod_reqtimeout erroneous timeouts
LP: #1561007 - (New)            24.03.16 [squid3]                       - Upstream Bug #3769: client_netmask not evaluated since Comm redesign
LP: #1375821 - (Triaged)        03.03.16 [ifupdown]                     - ifupdown doesn't bring loopback to state up
LP: #1267564 - (Confirmed)      21.10.15 [etckeeper]                    - etckeeper with git breaks package upgrades that touch /etc
LP: #1505556 - (Triaged)        16.10.15 [munin]                        - lpstat plugin doesn't return job count
LP: #1335391 - (Confirmed)      23.09.15 [etckeeper]                    - etckeeper with git breaks update-manager
LP: #1492212 - (Triaged)        04.09.15 [bind9]                        - postinst should validate config before restarting bind
LP: #1481337 - (Confirmed)      01.09.15 [keepalived]                   - keepalived makes a floating IP available on more than one host after configuration reload
LP: #1481579 - (New)            05.08.15 [amavisd-new]                  - amavisd-new throws stderr when running cron job
LP: #1242376 - (Triaged)        27.07.15 [php5]                         - reload php5-fpm crashes
LP: #1428796 - (New)            06.03.15 [apache2]                      - service stop; service start fails with wsgi.py from freeipa-server
LP: #609625  - (Confirmed)      14.05.11 [linux]                        - kexec-tools 0_kdump
LP: #28706   - (Confirmed)      17.08.10 [nfs-utils]                    - need way to specify the lockd port
LP: #150649  - (Confirmed)      14.06.10 [libapache-mod]                - gutsy does not have a working apache+mysql authentication solution
LP: #338047  - (Triaged)        12.03.10 [redhat-cluste]                - soft lockup - CPU#0 stuck for 11s! [clurgmgrd:4587]
LP: #222620  - (Confirmed)      10.08.08 [ebox-ntp]                     - package ebox-ntp None failed to install/upgrade: subprocess post-removal script returned error exit status 1
LP: #214863  - (Confirmed)      09.04.08 [nagios-plugin]                - various bugs and trust issues
```

Showing other tags than server-next now works properly
```
ustriage --show-no-triage --show-tagged --tag bitesize

Ubuntu Server Triage helper
Please be patient, this can take a few minutes...

---
Bugs tagged "bitesize" and subscribed "ubuntu-server"
Found 18 bugs
LP: #1531184 - +(Confirmed)     18.09.20 [dnsmasq]                      - [SRU] dnsmasq doesn't start on boot because its interface isn't up yet
LP: #1690836 - +(Confirmed)     27.06.18 [libnl3]                       - libnl-genl-3.0 memory leak
LP: #1705720 - +(Triaged)       22.04.20 [logwatch]                     - error message when run against amavisd:  Redundant argument in sprintf at /usr/share/logwatch/scripts/services/amavis line 1338
LP: #1755189 - +(Triaged)       12.03.19 [squid3]                       - squid3 helper basic_pop3_auth crashes
LP: #1772915 - +(Triaged)       24.06.21 [php8.0]                       - Documentation on PHP session garbage collection needs updating.
LP: #1792298 - +(Triaged)       16.06.21 [keepalived]                   - keepalived: MISC healthchecker's exit status is erroneously treated as a permanent error
LP: #1883469 - +(Triaged)       17.06.20 [sssd]           => hillpd     -  idmap_sss: improve man page
LP: #1909816 - +(Triaged)       17.03.21 [asterisk]                     - asterisk spams console output to syslog due to systemd misconfiguration
LP: #1919965 - +(Triaged)       20.09.21 [nginx]          => teward     - nginx-common nginx.service comments use wrong signal
LP: #782890  - +(Triaged)       22.11.18 [backuppc]                     - no ping response on all hosts after upgrade
LP: #1511222 - (Triaged)        02.08.16 [apache2]        => wesley-wi  - Incorrect trusted proxy match test in mod_remoteip
LP: #1524635 - (Triaged)        14.06.16 [haproxy]                      - haproxy syslog configuration causes double logging
LP: #1611923 - (Fix Committed)  29.09.16 [python-django]  => jgrimm     - http.request does not support ipv6-formatted ipv4 addresses
LP: #1728723 - (Triaged)        15.09.20 [postfix]                      - postfix check warns about symlink being group writable
LP: #1828074 - (In Progress)    21.09.21 [augeas]         => mirespace  - Memory usage issues with large Apache httpd configuration files
LP: #1882527 - (Triaged)        13.09.21 [mysql-5.7]      => paride     - mysql timeoutsec results in killing mysql process
also:        - (Fix Committed)           [mysql-8.0]      => paride    
LP: #1891643 - (Triaged)        12.06.21 [amavisd-new]                  - amavisd-new suggests missing packages
```

Showing all bugs of a person or team:
```
$ python -m ustriage --show-no-triage --show-subscribed --lpname paelzer
Ubuntu Server Triage helper
Please be patient, this can take a few minutes...

---
Bugs subscribed to paelzer
Found 311 bugs
LP: #1944204 - (New)            20.09.21 [landscape-cli]                - IP report in motd should ignore virtual interfaces
LP: #1749393 - (Triaged)        20.09.21 [qemu]                         - sbrk() not working under qemu-user with a PIE-compiled binary?
LP: #1906266 - (New)            15.09.21 [nova]                         - After upgrade: libvirt.libvirtError: Requested operation is not valid: format of backing image %s of image %s was not specified
LP: #1940957 - (Fix Committed)  13.09.21 [dpdk]                         - DPDK ports get disabled after Open vSwitch restart with Intel XXV710(i40e) and 25G AOC cables
LP: #1940913 - (Fix Committed)  13.09.21 [dpdk]                         - MRE updates 19.11.10 and 20.11.3
LP: #1941716 - (Triaged)        13.09.21 [needrestart]                  - Apt hook of needrestart  hang  with DEBIAN_FRONTEND=noninteractive
LP: #1940029 - (New)            04.09.21 [gcc-11]                       - Default of fcf-protection should only be enabled where it can work
LP: #1941719 - (New)            26.08.21 [grub2]                        - Please Recommend/Enhance qemu-system-x86
...
```


We still default to a subscription. Therefore you can not easily search for e.g. all bitesize no matter the subscription. But we might add --no-default-subscription one day if that is wanted?

paride commented:
Ping or re-request review from me when ready, and I'll have another look. Thanks for your work here, these changes will really help managing the bug queue!

paride reviewed (changes requested):
The request changes is mostly to mention the new options in `README.md`, but see my last comment in the conversation for this PR for more thoughts.

When using `--show-no-triage` the `+` symbol is used but never defined, e.g.:

```
Bugs tagged "server-next" and subscribed "ubuntu-server"
Found 13 bugs
LP: #1719354 - +(Triaged)
```
A minor thing, but the definition for `+` and `*` could be moved in the "common" printout section.

cpaelzer commented:
Discussion

> I still find the new --show-* options a bit "glued on" the tool, but in the end our aim for ustriage is to help with our process, not to write a "swiss army knife" for LP bug queries (which still it would be nice to have!), so I'm overall +1.

Yeah the swiss army knife will be more likely created in pdbq, I've talked with @bryceharrington about that. It isn't there yet but code-architecture wise that seems more extendable and a better place to eventually be our swiss army knife. But right now ustriage is closer to what we need and so "tucking on" a small extra use-case doesn't make the best code, but it isn't breaking it and gets us what we need in the short term.

> I hope this doesn't sound as "paride bikeshedding on option names again"

Not at all, while I really just want a quick solution to get our bug-housekeeping tool-assisted there is no point in making it worse than it could be as long as we do not tip over into the "perfection is the enemy of progress" trap. But we are still far away, so I'm absolutely happy about this.

> The bug lists have very long lines now, even in the default triage mode. Let's not change them in this PR, but we may think of reformatting them so they're more friendly with smaller terminal windows and copypasting in the triage report email.

Hmm, I think this becomes a matter of personal preference which I'm usually on the "give me everything" side of things and I have a lot of screen estate to throw chars at :-). But I see your POV, and want to adapt to make it useful for everyone.
The old format was:
`LP: #1403499 - *+(Fix Committed)    [apache2]        - systemd unable to detect running apache if invoked via apache2ctl graceful`
New:
`LP: #1832182 - *+(Fix Committed)  28.07.21 [apache2]        => bryce      - systemd unable to detect running apache if invoked via apache2ctl graceful`

I have already had added limiting to pkg-name and subscriber to make it less jagged on reports as pkg-name often exceeded the printf format. IMHO the old output already was sometimes even wider than my terminal, an example are just long titles.
Here in old format.
`LP: #1944454 - (Incomplete)     [clamav]         - package clamav-daemon 0.103.2+dfsg-0ubuntu0.20.04.2 failed to install/upgrade: installed clamav-daemon package post-installation script subprocess returned error exit status 1`

I think we have two options here, one is to length-limit the text and the other one is to block the new extra info (date/assignee) behind an option switch. Those are not mutually exclusive, so I'll try to do both.


Adapted:
- I clarified the argument as well as the help texts for tag & expiry
-  definition for + and * are in in the "common" printout section.
- show tagged/subscribed got an update to their help text clarifying that they add extra lists
- renamed --show-no-triage to --no-show-triage
- renamed --limit-subscribed to --show-subscribed-max
- added the missing ')'
- limited bug title to 60 chars
- hide extra info behind --extended-format (keeps the default layout more intact, and allows more for those who want)
- added matching entries to Readme.md

paride reviewed (approved):
LGTM and works fine. One cosmetic nit from `--help`:
```
  -T, --show-tagged     Display an additional list of bugs that --lpname or
                        "ubuntu-server") is directly subscribed to and are
                        tagged by (--tag or "server-next")
  -B, --show-subscribed
                        Display an additional list of bugs that (--lpname or
                        "ubuntu-server") is directly subscribed to
```
There's a missing `(` in the `-T` help text.

I'm approving this but not merging, in case you still want to tweak something or get a +1 from other team members before merging.

bryceharrington commented:
> > I still find the new --show-* options a bit "glued on" the tool, but in the end our aim for ustriage is to help with our process, not to write a "swiss army knife" for LP bug queries (which still it would be nice to have!), so I'm overall +1.
> 
> Yeah the swiss army knife will be more likely created in pdbq, I've talked with @bryceharrington about that. It isn't there yet but code-architecture wise that seems more extendable and a better place to eventually be our swiss army knife. But right now ustriage is closer to what we need and so "tucking on" a small extra use-case doesn't make the best code, but it isn't breaking it and gets us what we need in the short term.
> 

Yep, more general-purpose bug report querying is in the plans, but even with that it's going to be very helpful having well-thought out front end interfaces, so work towards perfecting ustriage's cli options is definitely not wasted effort.  :-)

One thing I'd add to the discussion is possibly incorporating a config file for per-user preferences.  With the cli tools included in pdbq I've been trying to ensure most options could alternatively be set in a YAML config file loaded from ~/.config.  It strikes me that some of the new --show-* options would be great candidates to be placed in such a config file, since they're almost certainly going to be preferred by some of us one way, and some of us another.
